draft-ietf-tls-ctls.md

```diff
@@ -595,6 +590,26 @@ This document requests that IANA open a new registry entitled "cTLS Template Key
 | finishedSize           | number       | (This document) |
 | optional               | object       | (This document) |
 
+## Reserved profiles
+
+This document requests that IANA open a new registry entitled "Well-known
+cTLS Profile IDs", on the Transport Layer Security (TLS) Parameters page,
+with the following columns:
+
+* ID value: A sequence of 1-4 octets.
+* Template: A JSON object.
+* Note: An explanation or reference.
+
+The ID value of length 1 are subject to a "Standards Action" registry
+policy. Values of length 2 are subject to an "RFC Required" policy. Other
+values are subject to a "First Come First Served" policy.
+
+The initial registry contents are:
+
+| ID value  | Template         | Note          |
+|:=========:|:================:|:=============:|
+| 0x00      | {"version": 772} | cTLS 1.3-only |
```

Comment:
are you sure the ID value is `0x00` and not an empty string?

Reply:
But 0x00 is not the empty string "" is.  Are strings NUL terminated?  https://github.com/tlswg/draft-ietf-tls-ctls/pull/55/files/b3ccd1a338de8d4a851cce8d7b5a26ccf4397683#diff-71b811173d0ff96753ecc5795f5a67be555770bcfe01c345ca9c1ed644e07306R151 shows a string inside quotes; this should match that I think.

Reply:
A profile ID is an octet sequence.  In JSON, this is represented as a hex-encoded string, but not all hex strings are valid (e.g. odd length strings) or distinct (e.g. capitalization).

I've added square brackets around the registered ID value to suggest that this is a list of length 1 containing the value zero.

Reply:
A list of length 1 is not a valid hex string.

Why not just make it a text string in netsascii or utf8?  Seems *much* simpler. I mean the cipher suites are strings...

Reply:
A profile ID is an octet sequence containing up to 255 octets.  The registry contains one such octet sequence (the sequence of length one whose first and only octet is zero).  This is all the same as ALPN, except that values longer than 4 are effectively "Private Use".

Specifying the profile ID as netascii seems dangerous, because JSON strings are always unicode.  It seems to me that an implementation could easily fail to handle some aspect of the netascii<->unicode conversion (e.g. embedded NUL bytes), leading to compatibility or security problems.

Specifying the profile ID as UTF-8 is inconvenient and inefficient for randomly generated profile IDs.  Instead of simply choosing N > 4 random octets, the profile generator has to do something more complicated (e.g. `profile_id = base64(random)`), and the resulting ClientHello will be larger for no reason (counter to the purpose of cTLS).

Reply:
Trying to optimize for randomly-generated profile IDs?  Is that really a common use-case? My conception was that there would be, maybe, a couple-dozen in widespread use.  But if you really need to do that, getting `n` random bytes and making them utf8 is essentially zero cost.

Reply:
cTLS profiles can contain certificates and even key shares, so they will need to be generated and rotated automatically.  If cTLS is in widespread use, there could easily be millions of distinct profiles.

The comparison to cipher suites seems relevant.  Cipher suites are identified by human-readable strings, but the cipher suite identifier transmitted on the wire is a compact binary value.  In this case, the `profile_id` is sent verbatim on the wire, so it is natural for it to also be a compact binary value.

Reply:
> cTLS profiles can contain certificates and even key shares, so they will need to be generated and rotated automatically. If cTLS is in widespread use, there could easily be millions of distinct profiles

That's not how I was thinking of things, but okay I withdraw the suggestion.